simd/unit_tests/include/TestSIMD_GeneratorCtors.hpp

```diff
@@ -36,20 +36,33 @@ inline void host_check_gen_ctor() {
     expected[i] = (init_mask[i]) ? init[i] * 9 : init[i];
   }
 
-  simd_type basic(KOKKOS_LAMBDA(std::size_t i) { return init[i]; });
-  mask_type mask(KOKKOS_LAMBDA(std::size_t i) { return init_mask[i]; });
-
   simd_type rhs;
   rhs.copy_from(init, Kokkos::Experimental::element_aligned_tag());
-  host_check_equality(basic, rhs, lanes);
-
-  simd_type lhs(KOKKOS_LAMBDA(std::size_t i) { return init[i] * 9; });
-  simd_type result(
-      KOKKOS_LAMBDA(std::size_t i) { return (mask[i]) ? lhs[i] : rhs[i]; });
 
   simd_type blend;
   blend.copy_from(expected, Kokkos::Experimental::element_aligned_tag());
-  host_check_equality(blend, result, lanes);
+
+  if constexpr (std::is_same_v<Abi, Kokkos::Experimental::simd_abi::scalar>) {
+    simd_type basic(KOKKOS_LAMBDA(std::size_t i) { return init[i]; });
+    host_check_equality(basic, rhs, lanes);
+
+    simd_type lhs(KOKKOS_LAMBDA(std::size_t i) { return init[i] * 9; });
+    mask_type mask(KOKKOS_LAMBDA(std::size_t i) { return init_mask[i]; });
+    simd_type result(
+        KOKKOS_LAMBDA(std::size_t i) { return (mask[i]) ? lhs[i] : rhs[i]; });
+
+    host_check_equality(blend, result, lanes);
+  } else {
+    simd_type basic([=](std::size_t i) { return init[i]; });
+    host_check_equality(basic, rhs, lanes);
+
+    simd_type lhs([=](std::size_t i) { return init[i] * 9; });
+    mask_type mask([=](std::size_t i) { return init_mask[i]; });
+    simd_type result(
+        [=](std::size_t i) { return (mask[i]) ? lhs[i] : rhs[i]; });
+
+    host_check_equality(blend, result, lanes);
+  }
```

Comment:
I'm guessing this separation was needed because `KOKKOS_LAMBDA` being `__host__ __device__` function in device enabled builds but is being called from host-only generator constructors (non-scalar simd backends)?

Reply:
Yes.

Reply:
Is it less disgusting if you do KOKKOS_IF_ON_HOST inside the KOKKOS_LAMBDA?

Reply:
IMHO, it's conceptionally correct (and consistent with the other constructors) for the generator constructors to be host-only.

Reply:
Please clarify what builds are broken without that change.
My concern is that this code become less readable/maintainable.

Reply:
All versions?

Reply:
> Is it less disgusting if you do KOKKOS_IF_ON_HOST inside the KOKKOS_LAMBDA?

Not clear how that would work. The problem is about handing a host device function to the generator isn't it?

Reply:
Lets go with this for now. I think the splitting is clean enough.